src/langchain_google_firestore/utility/document_converter.py

```diff
@@ -0,0 +1,179 @@
+# Copyright 2024 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+import ast
+
+from typing import (
+    TYPE_CHECKING,
+    List,
+)
+
+from langchain_core.documents import Document
+
+IMPORT_ERROR_MSG = (
+    "`firestore` package not found, please run `pip3 install google-cloud-firestore`"
+)
+
+if TYPE_CHECKING:
+    from google.cloud.firestore_v1.client import Client
+    from google.cloud.firestore_v1.base_document import DocumentSnapshot
+    from google.cloud.firestore_v1.document import DocumentReference
+
+
+class DocumentConverter:
+
+    @staticmethod
+    def convertFirestoreDocument(
+        document: DocumentSnapshot,
+        page_content_fields: List[str] = None,
+        metadata_fields: List[str] = None,
+    ) -> Document:
+        data_doc = document.to_dict()
+        metadata = {"reference": {"path": document.reference.path}}
+
+        if page_content_fields:
+            set_page_fields = set(page_content_fields)
+        else:
+            set_page_fields = set()
+        if metadata_fields:
+            set_metadata_fields = set(metadata_fields)
+        else:
+            set_metadata_fields = set()
+        shared_keys = set_metadata_fields & set_page_fields
+
+        page_content = {}
+        for k in sorted(shared_keys):
+            if k in data_doc:
+                val = DocumentConverter._convertFromFirestore(data_doc.pop(k))
+                page_content[k] = val
+                metadata[k] = val
+
+        metadata.update(
+            {
+                k: DocumentConverter._convertFromFirestore(data_doc.pop(k))
+                for k in sorted(set_metadata_fields - shared_keys)
+                if k in data_doc
+            }
+        )
+
+        if not set_page_fields:
+            # write all fields
+            keys = sorted(data_doc.keys())
+            page_content = {
+                k: DocumentConverter._convertFromFirestore(data_doc.pop(k))
+                for k in keys
+            }
+        else:
+            page_content.update(
+                {
+                    k: DocumentConverter._convertFromFirestore(data_doc.pop(k))
+                    for k in sorted(set_page_fields - shared_keys)
+                    if k in data_doc
+                }
+            )
```

Comment:
### About the `pop`
* With the `pop` I don't need to keep track of which keys I already assigned to `page_content` or `metadata`. 
* I can iterate once through the `metadata_fields` and remove those values so I don't keep track of the assigned keys and then add it to the `page_content`. If I had to keep track of the keys more conditionals are needed.
* When using `doc.to_dict()` we get a copy and not a reference so we are not modifying the original document.
* If `metadata_fields` is None/empty I need to set the remaining keys-values that where not assigned to `page_content`; removing the keys-values from the data is an alternative from tracking which keys I assigned and adding a conditional.
** A similar logic applies if `page_content_fields` is empty. I need to write the `metadata_fields`, keep track of the items and assign the remaining to `page_content`.

### For the sets (shared keys, and rest)
* If `page_content_fields` and `metadata_fields` share keys.  Both `page_content` and `metadata` will have the information.  
* The rest is to assign the remaining key-values. I don't like requiring the `sorted` for the testing and I'm reducing the number of elements before the sort. Otherwise we could iterate, with the `pop` we are sure we won't be adding duplicates.

* The behaviors when the metadata_fields or page_content_fields , and the testing flakiness led to this approach. Still there are similar iterations. I can iterate once for metadata_fields, pop the key-values and then go for `page_content`. But again if there are page_content_fields and metadata_fields is None. The remaining values unassigned go to `metadata`